File: modules/bar/bar.ts
import { Widget, Roundedges, App, Service } from "imports";
import options from "options";

const { RoundedAngleEnd } = Roundedges
const { Window, Box, CenterBox, Button, Icon } = Widget
const mpris = await Service.import("mpris");


// Widgets
import Workspaces from "./workspaces"
import { Title } from "./AppTitleTicker"
import { TickerBTN } from "./MediaTicker"
import { SysInfo } from "./sysinfo"
import { TrayReveal, Expandbtn } from "./tray"
import { Clock } from "./clock"

const pos = options.bar.position.bind()

const Dashbtn = () => Button({
	className: 'BarBTN',
	css: 'padding-right: 0.5rem;',
	tooltip_text: 'Dashboard',
	onClicked: () => App.toggleWindow("dashboard"),
	child: Icon({ icon: 'nix-snowflake-symbolic' }),
});

const Powerbtn = () => Button({
	className: 'BarBTN',
	css: 'padding-right: 0.5rem;',
	tooltip_text: 'Power Menu',
	child:
		Icon({
			icon: 'preferences-system-network-wakeonlan'
		}),
	onClicked: () => App.toggleWindow("sessioncontrols")
});

const Left = () => Box(
	{
		className: "barleft",
		hpack: "start",
		vpack: "center",
	},
	Workspaces(),
	RoundedAngleEnd("topright", { className: "angleRight" }),
	RoundedAngleEnd("topleft", { className: "angleLeft" }),
	Title(),
	RoundedAngleEnd("topright", { className: "angleRight" }),
);

const Center = () => Box(
	{
		className: "barcenter",
		hpack: "center",
		vpack: "center",
		hexpand: true,
	},
	RoundedAngleEnd("topleft", { className: "angleLeft" }),
	Clock(),
	RoundedAngleEnd("topright", { className: "angleRight" }),
	RoundedAngleEnd("topleft", { className: "angleLeft" }),
	Expandbtn(),
	TrayReveal(),
	RoundedAngleEnd("topright", { className: "angleRight" }),
	SysInfo(),
);

const Right = () => Box(
	{
		className: "barright",
		hpack: "end",
		vpack: "center",
		hexpand: true,
	},
	RoundedAngleEnd("topleft", { className: "angleLeft" }),
	TickerBTN(),
	RoundedAngleEnd("topright", { className: "angleRight" }),
	RoundedAngleEnd("topleft", { className: "angleLeft" }),
	Dashbtn(),
	Powerbtn(),
);

export const Bar = () => Window(
	{
		name: "bar",
		layer: 'top',
		anchor: pos.as(pos => [pos, "right", "left"]),
		exclusivity: "exclusive",
	},
	CenterBox({
		className: "bar",
		hexpand: true,
		vexpand: true,
		start_widget: Left(),
		center_widget: Center(),
		end_widget: Right(),
	}),
);

File: Rust/Modules/Bar/Bar.rs
r
use imports::{Widget, Roundedges, App, Service};
use options;

use std::future::Future;

const RoundedAngleEnd = Roundedges::RoundedAngleEnd;
const Window = Widget::Window;
const Box = Widget::Box;
const CenterBox = Widget::CenterBox;
const Button = Widget::Button;
const Icon = Widget::Icon;

// let mpris = Service::import("mpris").await;

// Widgets
mod Workspaces;
use Workspaces::Workspaces;
mod App_Title_Ticker;
use App_Title_Ticker::Title;
mod Media_Ticker;
use Media_Ticker::TickerBTN;
mod SysInfo;
use SysInfo::SysInfo;
mod SysTray;
use SysTray::{TrayReveal, Expandbtn};
mod Clock;
use Clock::Clock;

let pos = options::bar.position.bind();

fn Dashbtn() -> Widget {
    Button {
        class_name: "BarBTN".to_string(),
        css: "padding-right: 0.5rem;".to_string(),
        tooltip_text: "Dashboard".to_string(),
        on_clicked: || App::toggle_window("dashboard"),
        child: Icon { icon: "nix-snowflake-symbolic".to_string() },
    }
}

fn Powerbtn() -> Widget {
    Button {
        class_name: "BarBTN".to_string(),
        css: "padding-right: 0.5rem;".to_string(),
        tooltip_text: "Power Menu".to_string(),
        child: Icon { icon: "preferences-system-network-wakeonlan".to_string() },
        on_clicked: || App::toggle_window("sessioncontrols"),
    }
}

fn Left() -> Widget {
    Box {
        class_name: "barleft".to_string(),
        hpack: "start".to_string(),
        vpack: "center".to_string(),
        children: vec![
            Workspaces(),
            RoundedAngleEnd("topright", "angleRight".to_string()),
            RoundedAngleEnd("topleft", "angleLeft".to_string()),
            Title(),
            RoundedAngleEnd("topright", "angleRight".to_string()),
        ],
    }
}

fn Center() -> Widget {
    Box {
        class_name: "barcenter".to_string(),
        hpack: "center".to_string(),
        vpack: "center".to_string(),
        hexpand: true,
        children: vec![
            RoundedAngleEnd("topleft", "angleLeft".to_string()),
            Clock(),
            RoundedAngleEnd("topright", "angleRight".to_string()),
            RoundedAngleEnd("topleft", "angleLeft".to_string()),
            Expandbtn(),
            TrayReveal(),
            RoundedAngleEnd("topright", "angleRight".to_string()),
            SysInfo(),
        ],
    }
}

fn Right() -> Widget {
    Box {
        class_name: "barright".to_string(),
        hpack: "end".to_string(),
        vpack: "center".to_string(),
        hexpand: true,
        children: vec![
            RoundedAngleEnd("topleft", "angleLeft".to_string()),
            TickerBTN(),
            RoundedAngleEnd("topright", "angleRight".to_string()),
            RoundedAngleEnd("topleft", "angleLeft".to_string()),
            Dashbtn(),
            Powerbtn(),
        ],
    }
}

pub fn Bar() -> Widget {
    Window {
        name: "bar".to_string(),
        layer: "top".to_string(),
        anchor: pos.as(|pos| vec![pos, "right".to_string(), "left".to_string()]),
        exclusivity: "exclusive".to_string(),
        child: CenterBox {
            class_name: "bar".to_string(),
            hexpand: true,
            vexpand: true,
            start_widget: Left(),
            center_widget: Center(),
            end_widget: Right(),
        },
    }
}

